metadata/src/main/java/org/apache/kafka/timeline/SnapshottableHashTable.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.timeline;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+
+/**
+ * SnapshottableHashTable implements a hash table that supports creating point-in-time
+ * snapshots.  Each snapshot is immutable once it is created; the past cannot be changed.
+ * We handle divergences between the current state and historical state by copying a
+ * reference to elements that have been deleted or overwritten into the most recent
+ * snapshot tier.
+ *
+ * Note that there are no keys in SnapshottableHashTable, only values.  So it more similar
+ * to a hash set than a hash map.  The subclasses implement full-featured maps and sets
+ * using this class as a building block.
+ *
+ * Each snapshot tier contains a size and a hash table.  The size reflects the size at
+ * the time the snapshot was taken.  Note that, as an optimization, snapshot tiers will
+ * be null if they don't contain anything.  So for example, if snapshot 20 of Object O
```

Comment:
To be more precise, it's the snapshot tier's hashtable will be null, right?

Reply:
1. The HashTier could be null
2. or the HashTier could be non-null but the deltaTable could be null
3. or the HashTier could be non-null with a non-null delta table

You would end up in situation #2 if you deleted something but it was still present in a snapshot.  The HashTier would record the old size but not have any delta entries.